nni/mutable/mutable.py

```diff
@@ -0,0 +1,482 @@
+    :meth:`simplify` is the method to get the basic dimensions (type 1).
+    Algorithms then use the simplified space to run the search, generate the samples
+    (which are also in the format of the simplified space), and then,
+    :meth:`freeze` is the method to get the frozen version of the space with the sample.
+
+    For example::
+
+        >>> from nni.mutable import Discrete
+        >>> mutable = Discrete([2, 3]) + Discrete([5, 7])
+        >>> mutable
+        Discrete([2, 3], label='global_1') + Discrete([5, 7], label='global_2')
+        >>> mutable.simplify()
+        {'global_1': Discrete([2, 3], label='global_1'), 'global_2': Discrete([5, 7], label='global_2')}
+        >>> sample = {'global_1': 2, 'global_2': 7}
+        >>> mutable.freeze(sample)
+        9
+
+    In the example above, we create a new mutable that is the sum of
+    two existing variables (with :class:`MutableExpression`),
+    and then simplify it to get the basic dimensions.
+    The *sample* here is a dictionary of parameters.
+    It should have the exactly same keys as the simplified space,
+    and values are replaced with the sampled values.
+    The sample can be used in both :meth:`contains` and :meth:`freeze`.
+
+    * Use ``if mutable.contains(sample)`` to check whether a sample is valid.
+    * Use ``mutable.freeze(sample)`` to create a fixed version of the mutable.
+
+    One final note, :class:`Mutable` is designed to be framework agnostic.
+    It doesn't have any dependency on deep learning frameworks like PyTorch.
+    """
+
+    def freeze(self, sample: Sample) -> Any:
+        """Create a *frozen* (i.e., fixed) version of this mutable,
+        based on sample in the format of :meth:`simplify`.
+
+        For example, the frozen version of an integer variable is a constant.
+        The frozen version of a mathematical expression is an evaluated value.
+        The frozen version of a layer choice is a fixed layer.
+
+        Parameters
+        ----------
+        sample
+            The sample should be a dict, having the same keys as :meth:`simplify`.
+            The values of the dict are the choice of the corresponding mutable,
+            whose format varies depending on the specific mutable format.
+
+        Returns
+        -------
+        The frozen version of this mutable.
+
+        See Also
+        --------
+        LabeledMutable
+        """
+        raise NotImplementedError()
+
+    def contains(self, sample: Sample) -> SampleValidationError | None:
+        """Check whether sample is validly sampled from the mutable space.
+
+        Parameters
+        ----------
+        sample
+            See :meth:`freeze`.
+
+        Returns
+        -------
+        Optionally a :exc:`~nni.mutable.exception.SampleValidationError` if the sample is invalid.
+        """
+        raise NotImplementedError()
+
+    def leaf_mutables(self, is_leaf: Callable[[Mutable], bool]) -> Iterable[LabeledMutable]:
```

Comment:
Who provides the if_leaf callback?

Reply:
Search algorithm. Algorithm should provide `is_leaf` based on the types of mutables they can handle.

`leaf_mutables` is not considered as first-level API, but a method to override. `simplify` is.

Reply:
So in most cases algorithms call `simplify()` without argument, and if someone want do something special they call `leaf_mutables()` with a customized `is_leaf()`?

Reply:
Correct.